### DISCUSSION TITLE
How to implement themes for a multidomain application?

### DISCUSSION THREAD

**Discussion by kberov** [mojolicious/mojo#1630] [Q&A] [answered]
I have an application which serves several domains. 

Each incoming request may be for a different than the previous domain.
I would like to have different themes for the different domains - same templates names -  just in different renderer paths.

So far I implemented in `around_dispatch` to switch the template paths,
```perl
sub _around_dispatch ($next, $c) {
# ...
  state $s_paths = $app->static->paths;
  state $r_paths = $app->renderer->paths;
#...complex logic here
  unshift @{$s_paths}, "$root/$dom->{domain}/public";
  unshift @{$r_paths}, "$root/$dom->{domain}/templates";
  $next->();
  shift @{$s_paths};
  shift @{$r_paths};
}
```

But templates from different paths have the same names and Mojolicious::Renderer does not differentiate the cached templates by name.
On the next request(different domain from the previous one) in morbo I see:
```
[2020-12-18 23:23:20.34523] [9675] [debug] [8M98jmSZ] Rendering cached template "foo/bar.html.ep"
[2020-12-18 23:23:20.34569] [9675] [debug] [8M98jmSZ] Rendering cached template "partials/bazz.html.ep"
# etc...
```
How to set the template name so for `Mojolicious::Renderer` the templates
`/here/foo/bar.html.ep` and `/there/foo/bar.html.ep` are different.
Please note that  `/here` and `/there` are my template roots - different on each request.

**Comment by kberov** [comment]
This works for now, but only in CGI mode when the templates are not cached.

**Comment by Grinnz** [comment]
You can disable the template cache if you wish, see https://stackoverflow.com/a/41751645/5848200. This of course can be a significant performance impact as every request will have to read templates from disk.

Perhaps it would be a more efficient (and cacheable) setup to have the domains as subdirectories within the template path, prepended at the time of selecting the template.

**Comment by Grinnz** [comment, marked as answer]
Another more common idea is just to run multiple instances of the app - using Toadfarm or your reverse proxy to dispatch to the correct app based on domain, and having the template/static dirs read from config for each instance.

**Reply by kberov** [reply]
Thanks for the hints @Grinnz .
I would prefer if I have the option to influence the cache key for the template.

https://github.com/mojolicious/mojo/blob/master/lib/Mojolicious/Plugin/EPRenderer.pm#L20
```perl
# $renderer->template_name($options) gives different name if I add some additional option in before_render hook.
      my $name = $options->{inline} // $renderer->template_name($options);
      return unless defined $name;
      my $key = md5_sum encode 'UTF-8', $name;

      my $cache = $renderer->cache;
      my $mt    = $cache->get($key);
      $cache->set($key => $mt = Mojo::Template->new($ep)) unless $mt;
```
May be I can just make "my own" renderer with only this part slightly changed

**Reply by kberov** [reply]
@Grinnz, currently the application is on a cheap shared hosting, running as CGI (no caching of templates) and there is no problem, but if the time come when I have to move to a stand-alone app served by hypnotoad, I would prefer to be able to do it without changing the file/folders layout.

**Reply by Grinnz** [reply]
The caching mechanism is a Mojo::Cache object accessible via `$app->renderer->cache`, and thus can be replaced or customized as https://metacpan.org/pod/Mojolicious::Plugin::Renderer::WithoutCache does for example.

**Reply by mche** [reply]
>  cache key

Cache key would be full path to disk file instead the file name. So your put project folders _templates@mydomain1.com_, _templates@mydomain2.com_ ... and _static@mydomain1.com_, _static@mydomain2.com_ ...

And then around_dispatch

**Reply by kberov** [reply]
Thanks @mche.  I have different domain folders. Not sure how to implement your proposal(предложение).
I implemented `$cache->key_prefix($dom->{domain});` and used it in `around_dispatch`, so later (during the same request) when the renderer caches the template, found in the respective domain specific folder, it caches it with the specific domain key prefix. In my application the prefix and the renderer, and static paths are switched on each request, depending on the current domain. On the next request the renderer will find the cached template using the respective domain's prefix.

This is action at a distance. Otherwise I had to implement my own renderer, because I did not find another way to tell EPLRenderer what the [name](https://github.com/mojolicious/mojo/blob/4bf48eeea2a05ffe9ae70a53c85055d621c9582f/lib/Mojolicious/Plugin/EPLRenderer.pm#L12) of the template is.

May be I can get rid of my [Slovo::Cache](https://github.com/kberov/Slovo/blob/master/lib/Slovo/Cache.pm) if I do not need `cache_prefix`, but did not figured out how to do it.
And here is my current implementation of [around_dispatch](https://github.com/kberov/Slovo/blob/8f39fd46049df0b651707a3a815052660d786c0c/lib/Slovo.pm#L118).
Can you give a concrete idea with example? Thanks again.

**Reply by mche** [reply]
@kberov, I had own patched cache key from name to path at [Mojolicious::Plugin::EPRenderer](https://metacpan.org/release/Mojolicious/source/lib/Mojolicious/Plugin/EPRenderer.pm#L24)

`my $name = $options->{inline} // $renderer->template_path($options) || $renderer->template_name($options);`

I dont know why that is impossible global patch for team.
